### ISSUE TITLE
Add support for Show Cue System

### ISSUE THREAD

**Issue by Jus17** [stagehacks/Cue-View#57]
Hi, please add support to show cue system, it's the only real alternative to Qlab for windows. 
https://www.showcuesystems.com/cms/

The developer of the application is really active!

Thank you

**Comment by jwetzell** [comment]
I have yet to find any documentation on the TCP Remote App Interface that is mentioned in the documentation (still searching though). There does seem to be a mobile app made by someone so the interface is definitely capable of providing the information needed.

**Comment by Jus17** [comment]
Ok thank you, if you need I can contact the developer too for more informations

**Comment by jwetzell** [comment]
It seems to be just OSC, but I have yet to find any sort of documentation from SCS as to a "OSC Dictionary".

**Comment by Jus17** [comment]
Thank you Joel for your work

Il gio 9 feb 2023, 21:43 Joel Wetzell ***@***.***> ha scritto:

**Comment by jwetzell** [comment]
Surprisingly making some progress on SCS... 

<img width="583" alt="Screenshot_20230220_140751" src="https://user-images.githubusercontent.com/18341515/220190705-8eaa3930-4cfb-4d1e-b387-522b5d9ba885.png">
Hit some big blockers so waiting for further information from the Developer (and the SCSRemote developer) before we would be able to finish this and make it actually useful.

**Comment by Jus17** [comment]
Thank you!! Wonderful!

Il lun 20 feb 2023, 21:10 Joel Wetzell ***@***.***> ha
scritto:

**Comment by jwetzell** [comment]
Still haven't heard anything helpful from either of the developers, starting to think it might be a problem with developing against the Demo version of SCS.

**Comment by Jus17** [comment]
If you want I can try your beta with my professional version of Scs.

Il mar 4 apr 2023, 18:43 Joel Wetzell ***@***.***> ha scritto:

**Comment by jwetzell** [comment]
I don't have anything to test as they haven't gotten back to me about how it should work for me to be able to write anything. I've tested multiple things against the Demo version but it seems their implementation of OSC is "non-standard" to be nice..lol So until they get back with how things should work (or if the demo version is crippled in some way regarding the remote interface) I can't really do anything more than just guess and check which so far has yielded no results.

**Comment by Jus17** [comment]
Ok I've understood, thank you

Il mar 4 apr 2023, 18:52 Joel Wetzell ***@***.***> ha scritto:

**Comment by JDogHerman** [comment]
I sent a message to the developer of show cue systems and he sent this back to me. He also said he was available with further questions.

[SCS_Text_Only_OSC.pdf](https://github.com/stagehacks/Cue-View/files/14544809/SCS_Text_Only_OSC.pdf)

**Comment by Jus17** [comment]
Wow thank you for keep me in touch..
Can you implement this kind of message?

Il Sab 9 Mar 2024, 04:48 Justin Herman ***@***.***> ha
scritto:

**Comment by JDogHerman** [comment]
He also said it should work under the demo license and is willing to provide a temp pro license so @jwetzell isn't limited by the time limits of the demo. He also says he can support regular osc messages and not just text only. His email is... SCS Support • support@showcuesystems.com

![image](https://github.com/stagehacks/Cue-View/assets/1299282/1bfd607d-cfe4-43e5-ae4e-cefebc6d9f75)

**Comment by jwetzell** [comment]
I got the same document when I reached out to him about the protocol. I tried to get more information about the ACTUAL OSC interface not the "text only" interface described in that document but received no response from him or the other person he told me to reach out to but this was about a year ago now.

**Comment by jwetzell** [comment]
The documentation says that to communicate with it via OSC you just need to drop the leading underscore. If I do that ALL meaningful responses from SCS stop happening. If I use the "text only" interface and send `/_status` I get the expected response of `/_status 0`. If I instead send an OSC message to address  `/status` I get a reply of a 4 null bytes `0x00` followed by a comma (`,`). 

I could also not figure out how to receive any of the "SCS-Initiated Commands" which are really the only thing that is needed to finish this plugin implementation. The documentation does not state but he (Mike) said that SCS tries to initiate a connection on port number + 1 to send these messages but I have setup test scenarios to listen for return connection on all surrounding ports and did not receive a connection from SCS. 

The biggest things I would need to finish this plugin are
- why does sending an OSC message cause no response from SCS despite SCS recognizing the actions (`/ctrl/xxxx`)
- how does the SCS-Initiated messages part work I don't see any sort of incoming connections from SCS (TCP or UDP on surrounding ports)

**Comment by JDogHerman** [comment]
I reached out to the developer and he said he found your email from Feb of last year and said he would be reaching out to you (joel) get you the non-text ocs data on Monday (tomorrow)

I hope this helps.

**Comment by Jus17** [comment]
Ok thank you

Il Dom 10 Mar 2024, 15:15 Justin Herman ***@***.***> ha
scritto:

**Comment by jwetzell** [comment]
I hope so too! It's been like 80% done for a year as you can see in the screenshot earlier in this thread. It'd be nice to wrap it up.

**Comment by jwetzell** [comment]
Progress is still being made on this. He reached out to me on 4/1 with a "beta" version of SCS that has support for OSC. I have been working with him since to test it and make sure it has the things needed to be able to display the information in Cue View.

**Comment by Jus17** [comment]
Perfect thank you.. your program is perfect to keep an eye on all our
devices!

Il Mer 3 Apr 2024, 19:39 Joel Wetzell ***@***.***> ha scritto:

**Comment by Jus17** [comment]
Perfect Jason, thank you for your work! With the implementation of Scs I
will connect to all my collaborators and our work in theatre will keep
going better and better!

Thank you again, will you release a new version or the implementation of
scs with the new osc standard is already implemented?

Il Mer 18 Set 2024, 08:20 Jason ***@***.***> ha scritto:

**Comment by MaiJianxiang** [comment]
The released official version of SCS 11.10.3 has already implemented the Remote App interface conforming to the OSC standard.

**Comment by Jus17** [comment]
Ok, thank you

Il Ven 20 Set 2024, 05:07 Jason ***@***.***> ha scritto:

**Comment by jwetzell** [comment]
Yeah, I was waiting to hear back from the developer of SCS as he said he would notify me when these changes made it to a released version, but looks like it already has. We have a [branch](https://github.com/stagehacks/Cue-View/tree/plugin-show-cue-systems) that I've been slowly working on it does technically work (see examples above) but SCS does not really provide a good OSC interface to make cue loading/displaying/updating a quick task so it is pretty slow as of right now.

**Comment by Huckcity** [comment]
Chiming in from a related angle: [WinsLab](https://github.com/Huckcity/winslab) is a free, MIT-licensed QLab-style show control app that runs natively on Windows (and macOS) — audio/MIDI/OSC/network/script cues, multi-output audio, fades, groups with auto-advance. If what you want is a full Windows-native cue stack rather than SCS integration into Cue-View, it may already cover your use case: https://huckcity.github.io/winslab/

(Full disclosure: I'm the author — mentioning it here because this thread is exactly the "Windows show control" gap it was built for.)
